### PR TITLE
Show context results in result tab

### PR DESCRIPTION
**Pull request by mitiao**
See: https://progress.opensuse.org/issues/13920
Current tab view shows as below:

screenshot_20180514_114336

### CODE REVIEW

**Comment by Martchus** [comment]
* What speaks against putting this in an extra tab? But I suppose it is ok to mix it with the existing tab, too.
* When looking at old job's, wouldn't this prevent listing any previous jobs at all because only new ones are shown anymore (except a very high limit is specified)? Not sure whether going further back from an already old job is a use-case in practice.

**Comment by okurz** [comment]
The color is a good idea but isn't light blue often used for a selection? When I saw the screenshot at first I got confused and it looked like a user selected something.
What I had in mind is to also include the current job in the middle and mark that with bold dark lines for separation. Also as @Martchus mentioned think about the corner cases. Maybe an ellipsis is the way to go, show the next more recent jobs, then ... then the latest but make sure to always show an equal ratio of both more recent as well as previous. I did not check with implementation but what stopped me was that I did not want to query the database twice for previous *and* next jobs so I recommend to implement lazy loading for the whole tab first

**Comment by coolo** [comment]
I also find the background colour suboptimal. Having the last job part of the list sounds appealing too.
And I wouldn't worry about a gap between the last and the jobs before. Just mark the rows differently.

E.g. with ⇤ for latest,←for later, →for earlier jobs and • for current. Prepended to the result link

**Comment by coolo** [comment]
and +1 for lazy loading that tab

**Comment by Martchus** [comment]
According to [my last benchmarks](https://progress.opensuse.org/issues/25814#note-8), lazy loading would only be a slight improvement. However, this might change if you do more extensive database queries now. You can test this by adding `profiling_enabled = 1` to your openQA config. Results will be stored in the subdirectory `nytprof` of your openQA checkout.

**Comment by mitiao** [comment]
Rebased this with recent changes, the way of using DBIx::Class::ResultSource::View is a bit complicated, got the point from @coolo from irc that i would replace table view with it.

**Comment by coolo** [comment]
Next to that I find 'Results' as tab name a little confusing to new users. Because there are no results of that job in that tab. It was 'Previous results' before and I understand the change - but it makes a little too generic now.

How about 'Neighbours'?

**Comment by okurz** [comment]
> How about 'Neighbours'?

The [original ticket](https://progress.opensuse.org/issues/13920) mentions the word "Context" which I would prefer

**Comment by coolo** [comment]
Hmm, that makes sense to you - but does it to people exploring the interface? I find 'Context' heavily generic

**Comment by okurz** [comment]
Well, be explicit about what is shown and say "Next & previous results" as so far other "related", e.g. same scenario but on other archs, is not shown

**Comment by coolo** [comment]
So you're already preparing the next feature request? :)

**Comment by okurz** [comment]
I'm not sure. More "related" tests are likely to find on the overview page so for now I would not ask for a feature to show more "related" than just previous and next in the same scenario.

**Comment by mitiao** [comment]
So vote for 'Neighbours', 'Related results' and 'Next & previous results' :)

**Comment by Martchus** [comment]
We have enough space anyways, so I would say 'Next & previous results'.

File: assets/javascripts/job_results.js [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        html += '( ' + row.duration + ' )';
+    } else {
+        if (row.state == 'running' || row.state == 'scheduled') {
+            html += 'Not yet: ' + row.state;
+        }
+    }
+    return html;
+}
+
+$(document).ready(function(){
```
[the remark this answers is not in the document]
Comment (mitiao) [reply assets/javascripts/job_results.js]: Fixed

File: templates/test/job_results.html.ep [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+                                <li>Current job</li>
+                                <li>Previous jobs if have</li>
+                                <li>...</li>
+                                <li>...</li>
+                            </ul>
+                        </p>';
+                    %>
+                    %= help_popover 'Help for job results' => $content, undef, undef, 'top'
+                </span>
+                <button id="get_job_results" class="btn btn-success" type="button">
```
[the remark this answers is not in the document]
Comment (mitiao) [reply templates/test/job_results.html.ep]: Fixed

**Comment by mitiao** [comment]
I chose 'Next & previous results' and moved parameters binding into Jobs.pm

**Comment by coolo** [comment]
ok, much improved. But as it's lazy loaded, I seriously wonder if we shouldn't stick to load 100 neighbours and leave out the limit option at all - it just clutters the UI and 100 is still very quick

**Comment by mitiao** [comment]
i agree and will open a new ticket to remove the limit option

**Comment by okurz** [comment]
hm, can we still select more than 100 neighbours from the UI then?

**Comment by coolo** [comment]
well, how many more do you need? We can default to 100 next + 400 prev (the current maximum) just as well - it's paginaged

**Comment by Martchus** [comment]
To answer @okurz question: Yes, it is possible. I just tested locally with https://openqa.suse.de/tests/1640305#previous.

Note that the number of entries the table shows is a bit strange at the beginning but it actually makes sense. Eg. when limiting to 400, the table will show 404 entries in the mentioned job. That's because there are 4 next results (including the current job) and 400 previous results. So this limit is applied separately. That means we can go as far in the history as before.

**Comment by coolo** [comment]
Yes, but if we hard code to 400 and you go to a year old job, you will end with the current job in page 40 of 80. But I consider this a rare use case.

**Comment by okurz** [comment]
1. there are use cases where I need as many as possible which turned out to be higher than 400 sometimes. So previously I used the URL query parameter. If this is still possible and you do not want to remove that then it's fine.

2. Also I consider it important to have the address bar of the browser updated so that we can references these links elsewhere, e.g. in bug reports. That seems to have gone broken.

3. You seem to have removed the "link to latest" and also removed the test for this. Please bring this back. Seems we missed this in the review :(

**Comment by okurz** [comment]
I think this is not the same. I am not talking about showing the latest job at the time, which is a single job, but the URLs we also use in the bug reporting template, e.g. https://openqa.opensuse.org/tests/latest?machine=uefi-2G&arch=x86_64&flavor=Gnome-Live&test=gnome-live&version=3.28.2&distri=opensuse

**Comment by mitiao** [comment]
ok, i see that, i have opened a new enhancement ticket for this tab, we could talk there for your requests :)
https://progress.opensuse.org/issues/36307

**Comment by Martchus** [comment]
@okurz Note that the route `/tests/latest` hasn't been removed (only the link on the page).

**Comment by okurz** [comment]
Sure, but I would appreciate if that regression fix still goes in rather soon so that we do not break existing workflows of the test reviewers

**Comment by coolo** [comment]
Noted

**Pull request merged**